File: k-random-bean-validation/src/test/java/io/github/krandom/validation/PositiveOrZeroAnnotationHandlerTest.java
package io.github.krandom.validation;

import static org.assertj.core.api.Assertions.assertThat;

import io.github.krandom.KRandom;
import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import jakarta.validation.constraints.PositiveOrZero;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Set;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class PositiveOrZeroAnnotationHandlerTest {
  private KRandom kRandom;

  @BeforeEach
  void setUp() {
    kRandom = new KRandom();
  }

  @Test
  void generatedBeanShouldBeValidAccordingToValidationConstraints() {
    byte byteValue = 0;
    short shortValue = 0;
    int intValue = 0;
    long longValue = 0L;
    float floatValue = 0F;

    TestBean testBean = kRandom.nextObject(TestBean.class);

    assertThat(testBean.getTestBigDecimal()).isGreaterThanOrEqualTo(BigDecimal.ZERO);
    assertThat(testBean.getTestBigInteger()).isGreaterThanOrEqualTo(BigInteger.ZERO);
    assertThat(testBean.getTestPrimitiveByte()).isGreaterThanOrEqualTo(byteValue);
    assertThat(testBean.getTestWrapperByte()).isGreaterThanOrEqualTo(byteValue);
    assertThat(testBean.getTestPrimitiveShort()).isGreaterThanOrEqualTo(shortValue);
    assertThat(testBean.getTestWrapperShort()).isGreaterThanOrEqualTo(shortValue);
    assertThat(testBean.getTestPrimitiveInt()).isGreaterThanOrEqualTo(intValue);
    assertThat(testBean.getTestWrapperInteger()).isGreaterThanOrEqualTo(intValue);
    assertThat(testBean.getTestPrimitiveLong()).isGreaterThanOrEqualTo(longValue);
    assertThat(testBean.getTestWrapperLong()).isGreaterThanOrEqualTo(longValue);
    assertThat(testBean.getTestPrimitiveFloat()).isGreaterThanOrEqualTo(floatValue);
    assertThat(testBean.getTestWrapperFloat()).isGreaterThanOrEqualTo(floatValue);
    assertThat(testBean.getTestPrimitiveDouble()).isGreaterThanOrEqualTo(0D);
    assertThat(testBean.getTestWrapperDouble()).isGreaterThanOrEqualTo(0D);
  }

  @Test
  void generatedBeanShouldBeValidUsingBeanValidationApi() {
    TestBean testBean = kRandom.nextObject(TestBean.class);

    Validator validator;
    try (ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory()) {
      validator = validatorFactory.getValidator();
    }
    Set<ConstraintViolation<TestBean>> violations = validator.validate(testBean);

    assertThat(violations).isEmpty();
  }

  static class TestBean {
    public BigDecimal getTestBigDecimal() {
      return testBigDecimal;
    }

    public BigInteger getTestBigInteger() {
      return testBigInteger;
    }

    public byte getTestPrimitiveByte() {
      return testPrimitiveByte;
    }

    public Byte getTestWrapperByte() {
      return testWrapperByte;
    }

    public short getTestPrimitiveShort() {
      return testPrimitiveShort;
    }

    public Short getTestWrapperShort() {
      return testWrapperShort;
    }

    public int getTestPrimitiveInt() {
      return testPrimitiveInt;
    }

    public Integer getTestWrapperInteger() {
      return testWrapperInteger;
    }

    public long getTestPrimitiveLong() {
      return testPrimitiveLong;
    }

    public Long getTestWrapperLong() {
      return testWrapperLong;
    }

    public float getTestPrimitiveFloat() {
      return testPrimitiveFloat;
    }

    public Float getTestWrapperFloat() {
      return testWrapperFloat;
    }

    public double getTestPrimitiveDouble() {
      return testPrimitiveDouble;
    }

    public Double getTestWrapperDouble() {
      return testWrapperDouble;
    }

    @PositiveOrZero private BigDecimal testBigDecimal;

    @PositiveOrZero private BigInteger testBigInteger;

    @PositiveOrZero private byte testPrimitiveByte;

    @PositiveOrZero private Byte testWrapperByte;

    @PositiveOrZero private short testPrimitiveShort;

    @PositiveOrZero private Short testWrapperShort;

    @PositiveOrZero private int testPrimitiveInt;

    @PositiveOrZero private Integer testWrapperInteger;

    @PositiveOrZero private long testPrimitiveLong;

    @PositiveOrZero private Long testWrapperLong;

    @PositiveOrZero private float testPrimitiveFloat;

    @PositiveOrZero private Float testWrapperFloat;

    @PositiveOrZero private double testPrimitiveDouble;

    @PositiveOrZero private Double testWrapperDouble;
  }
}

File: bean-validation/src/test/kotlin/io/github/krandom/validation/PositiveOrZeroAnnotationHandlerTest.kt
package io.github.krandom.validation

import io.github.krandom.KRandom
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.comparables.shouldBeGreaterThanOrEqualTo
import jakarta.validation.Validation
import jakarta.validation.Validator
import jakarta.validation.constraints.PositiveOrZero
import java.math.BigDecimal
import java.math.BigInteger
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class PositiveOrZeroAnnotationHandlerTest {
  private lateinit var kRandom: KRandom

  @BeforeEach
  fun setUp() {
    kRandom = KRandom()
  }

  @Test
  fun `generated bean should be valid according to validation constraints`() {
    val byteValue: Byte = 0
    val shortValue: Short = 0
    val intValue = 0
    val longValue = 0L
    val floatValue = 0F

    val testBean = kRandom.nextObject(TestBean::class.java)

    testBean.testBigDecimal shouldBeGreaterThanOrEqualTo BigDecimal.ZERO
    testBean.testBigInteger shouldBeGreaterThanOrEqualTo BigInteger.ZERO
    testBean.testByte shouldBeGreaterThanOrEqualTo byteValue
    testBean.testShort shouldBeGreaterThanOrEqualTo shortValue
    testBean.testInt shouldBeGreaterThanOrEqualTo intValue
    testBean.testLong shouldBeGreaterThanOrEqualTo longValue
    testBean.testFloat shouldBeGreaterThanOrEqualTo floatValue
    testBean.testDouble shouldBeGreaterThanOrEqualTo 0.0
  }

  @Test
  fun `generated bean should be valid using bean validation api`() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    val validator: Validator
    Validation.buildDefaultValidatorFactory().use { validatorFactory ->
      validator = validatorFactory.validator
    }
    val violations = validator.validate(testBean)

    violations.shouldBeEmpty()
  }

  internal data class TestBean(
    @field:PositiveOrZero val testBigDecimal: BigDecimal,
    @field:PositiveOrZero val testBigInteger: BigInteger,
    @field:PositiveOrZero val testByte: Byte,
    @field:PositiveOrZero val testShort: Short,
    @field:PositiveOrZero val testInt: Int,
    @field:PositiveOrZero val testLong: Long,
    @field:PositiveOrZero val testFloat: Float,
    @field:PositiveOrZero val testDouble: Double,
  )
}
